R/tostring.R

```diff
@@ -805,14 +1087,159 @@ wrap_string <- function(str, width, collapse = NULL) {
+      remainder = substr(str, length(ok) + 1, nchar(str))
+    )
+  }
+}
+
+## need a separate path here because **the number of characters**
+## in each part is no longer going to be constant the way it
+## was for monospace
+## this is much slower but still shouldn't be a bottleneck, if it is we'll
+## have to do something else
+#' wrap string given a Truetype font
+#'
+#' @inheritParams wrap_string
+#' @param min_ok_chars numeric(1). Number of minimum characters which much remain
+#' on either side when a word is split
+#' @param wordbreak_ok logical(1). Should breaking within a word be allowed. If, `FALSE`,
+#' attempts to wrap a string to a width narrower than its widest word will result
+#' in an error.
+#' @return `str`, broken up into a word-wrapped vector
+#' @export
+wrap_string_ttype <- function(str,
+                              width,
+                              fontspec,
+                              collapse = NULL,
+                              min_ok_chars = min(floor(nchar(str) / 2), 4, floor(width / 2)),
+                              wordbreak_ok = TRUE) {
+  newdev <- open_font_dev(fontspec)
+  if (newdev) {
+    on.exit(close_font_dev())
+  }
+
+  rawspls <- strsplit(str, "[[:space:]](?=[^[:space:]])", perl = TRUE)[[1]] # preserve all but one space
+  nctt <- nchar_ttype(rawspls, fontspec, raw = TRUE)
+  pts <- which(cumsum(nctt) <= width)
+  if (length(pts) == length(rawspls)) { ## no splitting needed
+    return(str)
+  } else if (length(pts) == 0) { ## no spaces, all one word, split it and keep going
+    if (wordbreak_ok) {
+      inner_res <- list()
+      min_ok_inner <- min_ok_chars
+      while (length(inner_res$ok) == 0) {
+        inner_res <- split_word_ttype(rawspls[1], width, fontspec, min_ok_inner) # min_ok_chars)
+        min_ok_inner <- floor(min_ok_inner / 2)
+      }
+      done <- inner_res$ok
+      remainder <- paste(c(inner_res$remainder, rawspls[-1]), collapse = " ")
+    } else {
+      stop(
+        "Unable to find word wrapping solution without breaking word: ",
+        rawspls[[1]], " [requires  ", nchar_ttype(rawspls[[1]], fontspec), " spaces of width, out of ",
+        width, " available]."
+      )
+    }
+  } else { ## some words fit, and some words don't
+    done_tmp <- paste(rawspls[pts], collapse = " ")
+    tospl_tmp <- rawspls[length(pts) + 1]
+    width_tmp <- width - sum(nctt[pts])
+    if (wordbreak_ok && width_tmp / width > .33) {
+      inner_res <- split_word_ttype(tospl_tmp, width_tmp, fontspec,
+        min_ok_chars = min_ok_chars
+      )
+    } else {
+      inner_res <- list(done = "", remainder = tospl_tmp)
+    }
+    done <- paste(c(rawspls[pts], inner_res$ok),
+      collapse = " "
+    )
+    remainder <- paste(
+      c(
+        inner_res$remainder,
+        if (length(rawspls) > length(pts) + 1) tail(rawspls, -(length(pts) + 1))
+      ),
+      collapse = " "
+    )
+  }
+  ret <- c(
+    done,
+    wrap_string_ttype(remainder, width, fontspec)
```

Comment:
eheh at the end I also went for a recursive solution. I did not expect it but it is incredibly faster too